### PR TITLE
Add SChannel TLS 1.3 support

### PR DESCRIPTION
**Pull request by wyattoday**
Add SChannel TLS 1.3 support to curl / libcurl.

Has detection at runtime to enable / disable support depending on version of Windows.

Also, added a new define to disable optional code: `DISABLE_SCHANNEL_CLIENT_CERT` (reduces compiled code size if that feature is not needed).

A bit more work needs to be done before this is "merge-ready", but TLS 1.3 support is working.

---

This depends on this PR being merged: https://github.com/curl/curl/pull/7727

Remaining work: Implement "set_ssl_ciphers" for `SCH_CREDENTIALS`. This is in the inverse of the `SCHANNEL_CRED` behavior.

Namely, we're disabling ciphers rather than enabling them. This will require a bit of thinking. I'll work on this this week, but if anyone wants to jump in and write that function, I'll be happy to accept the help.

Once that is implemented, everything else should be good to go. But I'll happily accept input on any of it.

### CODE REVIEW

File: lib/vtls/schannel.c [hunk 1]
```diff
@@ -706,27 +726,143 @@ schannel_acquire_credential_handle(struct Curl_easy *data,
...
-    s_pSecFn->AcquireCredentialsHandle(NULL, (TCHAR *)UNISP_NAME,
-                                       SECPKG_CRED_OUTBOUND, NULL,
-                                       &schannel_cred, NULL, NULL,
-                                       &BACKEND->cred->cred_handle,
-                                       &BACKEND->cred->time_stamp);
+  /* Windows 10, 1809 (a.k.a. Windows 10 build 17763) */
+  if(curlx_verify_windows_version(10, 0, 17763, PLATFORM_WINNT,
+      VERSION_GREATER_THAN_EQUAL)) {
+
+    SCH_CREDENTIALS Credentials = { 0 };
```
Comment (mback2k) [inline lib/vtls/schannel.c:746]: We use normally use lower-case names for local variables.

File: lib/vtls/schannel.c [hunk 2]
```diff
@@ -1974,7 +2112,7 @@ schannel_recv(struct Curl_easy *data, int sockindex,
           infof(data, "schannel: can't renegotiate, an error is pending");
           goto cleanup;
         }
-        if(BACKEND->encdata_offset) {
+        if(BACKEND->encdata_offset && inbuf[1].cbBuffer) {
```
Comment (mback2k) [inline lib/vtls/schannel.c:2115]: This change seems unrelated to TLS 1.3.

File: lib/vtls/schannel.c [hunk 3]
```diff
@@ -2049,15 +2187,20 @@ schannel_recv(struct Curl_easy *data, int sockindex,
   */
   if(len && !BACKEND->decdata_offset && BACKEND->recv_connection_closed &&
      !BACKEND->recv_sspi_close_notify) {
-    bool isWin2k = curlx_verify_windows_version(5, 0, PLATFORM_WINNT,
+
+#if defined(_WIN32_WINNT) && _WIN32_WINNT <= 0x0500
```
Comment (mback2k) [inline lib/vtls/schannel.c:2191]: I do not understand the whole thing of combining runtime and compile time checks like this. What is the purpose here?

File: lib/vtls/schannel.c (hunk 2 above)
Comment (wyattoday) [reply lib/vtls/schannel.c:2115]: It is required. I spend a few days mulling over and debugging what exactly was happening. With the new SChannel support of `SCH_CREDENTIALS`, inbuf[1] can be empty while `encdata_offset` is not empty when receiving a `SEC_I_RENEGOTIATE` return code.

Unfortunately, MS's documentation on this behavior is slim, so I can't tell you *why* this is the case, just that it *is* the case.

Long story short: without this change `SCH_CREDENTIALS` support (and thus TLS 1.3 support) will fail.

File: lib/vtls/schannel.c (hunk 3 above)
Comment (wyattoday) [reply lib/vtls/schannel.c:2191]: Removing dead code-paths (smaller compiler binary). This is inherited from the other PR that this PR depends on: https://github.com/curl/curl/pull/7727

File: lib/vtls/schannel.c (hunk 1 above)
Comment (wyattoday) [reply lib/vtls/schannel.c:746]: I'll fix those tomorrow, thanks.

Comment (wyattoday) [reply lib/vtls/schannel.c]: Never mind, had a free minute. Fixed now.

**Comment by gvollant** [comment]
Hello @wyattoday , I suggest you rebase and do a git fixup / squash to merge all (or somes of) your commit

File: lib/vtls/schannel.c (hunk 2 above)
Comment (mback2k) [reply lib/vtls/schannel.c:2115]: Understood, but what happens for applications not using TLS 1.3 following the old code path?

File: lib/vtls/schannel.c (hunk 3 above)
Comment (mback2k) [reply lib/vtls/schannel.c:2191]: I don't think this is worth the effort as it over-complicates the code and makes reading it a lot harder. Generally speaking we try to perform Windows specific checks during runtime, because it is common for a Windows application to be build on a one platform version and then shipped to a couple of others. Therefore compile-time checks do not make much sense with the Windows ecosystem.

File: lib/vtls/schannel.h [hunk 4]
```diff
@@ -61,16 +71,79 @@ CURLcode Curl_verify_certificate(struct Curl_easy *data,
 
 #ifdef __MINGW32__
 #include <_mingw.h>
-#ifdef __MINGW64_VERSION_MAJOR
+#if defined(__MINGW64_VERSION_MAJOR) && !defined(DISABLE_SCHANNEL_MANUAL_VERIFY)
```
Comment (mback2k) [inline lib/vtls/schannel.h:74]: I think the two new flags/options `DISABLE_SCHANNEL_MANUAL_VERIFY` and `DISABLE_SCHANNEL_CLIENT_CERT` should be done in a separate PR as this is unrelated to TLS 1.3.

> (SChannel already handles verification internally)

Also this is not true for all versions of Windows.

File: lib/version_win32.c [hunk 5]
```diff
@@ -31,6 +31,29 @@
...
+#define OLDER_THAN_WIN2K 0
+#endif
+
+/* Include the RtlGetVersion() regardless of the presence of
+   the Windows Driver Development Kit.
+*/
+#if !OLDER_THAN_WIN2K
+#include <winnt.h>
+#ifdef _MSC_VER
+#pragma comment(lib, "ntdll.lib")
```
Comment (mback2k) [inline lib/version_win32.c:47]: Thanks for working on this PR, but I guess reviewing it will be a lot easier once #7727 is settled. Not necessarily merged as it is right now as I don't think the discussion about ntdll is finished.

File: lib/vtls/schannel.c (hunk 2 above)
Comment (wyattoday) [reply lib/vtls/schannel.c:2113]: Same working behavior when using TLS 1.2 (or older) on either the `SCH_CREDENTIALS` or `SCHANNEL_CRED` code-paths.

File: lib/version_win32.c (hunk 5 above)
Comment (wyattoday) [reply lib/version_win32.c:47]: I agree. That's why I made the other PR a couple of weeks ago, so there would be ample time to get that merged and then this reviewed.

Unfortunately we needed TLS 1.3 support on Windows and we couldn't wait any longer. We figured it would be better to share what we had immediately since Windows 11 will be released next week. And we didn't want other people wasting their time re-implementing the work we already did.

File: lib/vtls/schannel.h (hunk 4 above)
Comment (wyattoday) [reply lib/vtls/schannel.h]: > Also this is not true for all versions of Windows.

Really? From my research (again, documentation is spotty on MSDN), auto-verification has been supported since at least Windows 2000.

Either way, I can make a separate PR that adds those 2 options. Or better yet, adds the inverse of `DISABLE_SCHANNEL_MANUAL_VERIFY` (`ENABLE_SCHANNEL_MANUAL_VERIFY`) so that by-default end-users will be smaller binaries.

Would that ease the merging of this PR? If so, I'll do that.

Comment (mback2k) [reply lib/vtls/schannel.h]: See the commit originally introducing it: https://github.com/curl/curl/commit/1e4c57fa6471c0010f66d82a83f190a07fa37309

**Comment by mback2k** [comment]
It seems like this is currently causing a lot of tests to fail on the various Windows CIs. This is definitely a blocker and needs to be fixed before we can consider this PR.

File: lib/vtls/schannel.h (hunk 4 above)
Comment (wyattoday) [reply lib/vtls/schannel.h]: Ah, so it was added for Windows CE. So, it makes sense to just enable manual verification for Windows CE only. So, do away with `DISABLE_SCHANNEL_MANUAL_VERIFY` and just use `_WIN32_WCE` (which looks like the original behavior implemented).

Comment (wyattoday) [reply lib/vtls/schannel.h]: OK, I've removed `DISABLE_SCHANNEL_MANUAL_VERIFY` and used `_WIN32_WCE` as was originally written.

File: lib/vtls/schannel.c (hunk 2 above)
Comment (jay) [reply lib/vtls/schannel.c:2113]: It looks like this change effectively renders the check moot. I added the check in 3e7ec1e, but going through it just now I can't remember why. Maybe I thought it was not possible that encrypted data would be sent before the renegotiate completed? However my note in the commit message contradicts that guess:

~~~
- Do not recurse on renegotiation.

Instead we'll continue on to process any trailing encrypted data
received during the renegotiation only.
~~~

So I think my intention was to process that data.

Comment (wyattoday) [reply lib/vtls/schannel.c:2113]: Unfortunately I couldn't reproduce a case where `SEC_I_RENEGOTIATE` was received, `BACKEND->encdata_offset` was `!= 0` and it wasn't just faulty error handling. The code *still* handle the case where `inbuf[1].cbBuffer == 0`  and `inbuf[3].cbBuffer != 0`. But is that necessary (or is it a code-path that will never be executed)? I have no idea -- I couldn't reproduce a case that got that result.

Unfortunately that commit message and those PR discussions didn't get me any closer to reproducing a test-case.

**Comment by vszakats** [comment]
To have a clearer picture of what this PR does, I'd suggest rebasing it on @jay's https://github.com/curl/curl/pull/7810, and thus dropping all version-detection-related extra bits from this patch.

It'd be good to see why `SCH_CREDENTIALS` is necessary to enable TLS 1.3. "Normally" a client isn't required to do anything extra to use a newer TLS version, if the underlying TLS-layer starts supporting it. [Microsoft says](https://docs.microsoft.com/windows/win32/secauthn/protocols-in-tls-ssl--schannel-ssp-) that TLS 1.3 should be supported by SChannel without extra settings. Anything in the contrary is possible of course, but the point is to only add code (esp. platform/version specific code), if absolutely necessary.

`Winternl.h` should lose the capitalization, otherwise it will not be found in cross-build situations where the build host uses a case-sensitive file system (e.g. Linux), where this header is distributed in full lowercase (i.e. mingw/WINE).

**Comment by wyattoday** [comment]
@vszakats 

> To have a clearer picture of what this PR does, I'd suggest rebasing it on @jay's #7810, and thus dropping all version-detection-related extra bits from this patch.

I will once that PR has landed and when I (or someone else) lands a PR adding build version checking.

> Microsoft says that TLS 1.3 should be supported by SChannel without extra settings.

They don't really say that in that table, though. Unfortunately as you've likely seen the documentation is sparse. I remember reading it either in a Microsoft blog by a network-stack programmer or a comment on that blog by said programmer.

Either way, you can confirm it yourself by passing the TLS_1_3 flag to the old `SCHANNEL_CRED` struct. You'll see a failure at the negotiation stage.

> Winternl.h should lose the capitalization, otherwise it will not be found in cross-build situations where the build host uses a case-sensitive file system (e.g. Linux), where this header is distributed in full lowercase (i.e. mingw/WINE).

I'll fix that once the rebasing on the fixed version checking is done. Thanks!

**Comment by bagder** [comment]
What's the current status of this work? It needs a rebase and force-push at the very least...

**Comment by wyattoday** [comment]
I was planning on revisiting this Thursday or Friday this week. I'll start with a rebase on main ("master"), and then make the handful of changes requested throughout the thread.

**Comment by gvollant** [comment]
> I was planning on revisiting this Thursday or Friday this week. I'll start with a rebase on main ("master"), and then make the handful of changes requested throughout the thread.

Yes, please do a fast rebase and merge commit. This will allow review

**Comment by wyattoday** [comment]
Close in favor of the new PR with completed work, build on "main".

https://github.com/curl/curl/pull/8419

**Comment by gvollant** [comment]
after merging #8419, the next step is pehaps support msquic on schannel , see https://github.com/microsoft/msquic/blob/main/docs/Platforms.md

**Comment by bagder** [comment]
@nibanks has started working on adding support for msh3 to curl (https://github.com/nibanks/curl/pull/1). msh3 uses mquic, which should be an excellent step torwards getting h3 support in curl on Windows powered by Schannel.

**Comment by Jrarmyjr** [comment]
union

**Comment by Jrarmyjr** [comment]
union
